app/server/utils/transfer_limits.py

```diff
@@ -157,160 +151,577 @@ def is_any_token_and_user_is_kyc_verified(credit_transfer):
 def is_any_token_and_user_is_kyc_business_verified(credit_transfer):
     return is_user_and_any_token(credit_transfer) and user_business_or_multidoc_kyc_verified(credit_transfer)
 
-# ~~~~~~LIMIT FILTERS~~~~~~
 
+# ~~~~~~LIMIT FILTERS~~~~~~
 
-@curry
-def after_time_period_filter(days: int, query: Query):
 
+def after_time_period_filter(days: int):
     epoch = datetime.datetime.today() - datetime.timedelta(days=days)
+    return [CreditTransfer.created >= epoch]
+
+
+def matching_sender_user_filter(transfer: CreditTransfer):
+    return [CreditTransfer.sender_user == transfer.sender_user]
+
+
+def regular_payment_filter(transfer: CreditTransfer):
+    return [CreditTransfer.transfer_subtype == TransferSubTypeEnum.STANDARD]
+
+
+def matching_transfer_type_filter(transfer: CreditTransfer):
+    return [CreditTransfer.transfer_type == transfer.transfer_type]
+
+
+def matching_transfer_type_and_subtype_filter(transfer: CreditTransfer):
+    return [
+        CreditTransfer.transfer_type == transfer.transfer_type,
+        CreditTransfer.transfer_subtype == transfer.transfer_subtype
+    ]
+
+
+def withdrawal_or_agent_out_and_not_excluded_filter(transfer: CreditTransfer):
+    return [
+        or_(CreditTransfer.transfer_type == TransferTypeEnum.WITHDRAWAL,
+            CreditTransfer.transfer_subtype == TransferSubTypeEnum.AGENT_OUT),
+        CreditTransfer.exclude_from_limit_calcs == False
+    ]
+
+
+def not_rejected_filter():
+    return [CreditTransfer.transfer_status != TransferStatusEnum.REJECTED]
+
+
+class BaseTransferLimit(ABC):
+    """
+    Base Limit Class. All limits use `applies_to_transfer` to determine if they are applied.
+    Specific Limit rules vary hugely, so this class is pretty sparse,
+    however, all limits follow the same overall process:
+    1. Check if the limit applies to the transfer, based off the a) Type/Subtype & b) Query Filters
+    2. Calculate how large the limit is (may be fixed, or variable)
+    3. Calculate how much of the limit is available (this may involve aggregating across prev transfers and subtracting)
+    4. Calculate how much the current transfer cases uses
+    5. Check if the current case uses more than what's available
+    """
+
+    @abstractmethod
+    def available(self, transfer: CreditTransfer) -> AggregateAvailability:
```

Comment:
I agree that there should definitely be no side effects, but I don't think Alchemy Objects should be abstracted out of this, because:
- a core part of the limits is constructing the appropriate database queries, so as far as i can tell, abstracting out those queries would just require them to be reconstructed elsewhere, which defeats the purpose
- Equally, it's impossible to fully define what methods and accessors may be needed for a query without loosing the extensibility of the limits. For example there's no limit that currently requires the Email of the sender to be set, but adding such a case would be trivial

That being said, open to example snippets of how this might work nicely!

Reply:
Also, (out of scope for this PR) I think putting some expectations on when out-of-system side effects occur when working with Alchemy Models is a good idea, though I don't think it's a good idea to enforce this as a hard rule. The gold standard of flask apps has side effects in its alchemy model:
https://github.com/miguelgrinberg/microblog/blob/v0.22/app/models.py

Reply:
> it's impossible to fully define what methods and accessors may be needed for a query without loosing the extensibility of the limits. 

without having the full picture here; you are already passing `CreditTransfer` here. Can't you replace that with an abstraction, to which you add - if need be - all the methods in the model object? Although, I would imagine that there should be a way of making more flexible accessors, for example through a key/value interface (`fcntl` style). This way, if the `CreditTransfer` cannot be instatiated without some magic happening, the component can be wrapped or replaced.

> The gold standard of flask apps has side effects in its alchemy model

Yes, I realize that. I guess what I am uncomfortable with here are canonical object representations in the system that _cannot_ even be _instantiated_ without having an active db (or blockchain) backend. If that can be solved in _all_ cases, I have no problem with directly using them in function signatures.

> open to example snippets of how this might work nicely

I'm sorry, I don't have the bandwidth. I would suspect you yourselves know better than me where any spurious side-effects occur, though...

Reply:
On further investigation, I don't think this is even possible - the Alchemy query filters eg `query.filter_by(CreditTransfer.id == foo)` require full alchemy objects to operate, so u can't define an interface that isn't reliant on Alchemy.

I can see the arguments for why the models shouldn't rely on db connections, but the entire app is effectively a CRUD app build around Alchemy models - if we wanted to remove that dependency it'd basically mean re-writing this from the ground up without Alchemy.